### PR TITLE
!syscalls: change mmap/munmap API to allow setting errno

### PR DESCRIPTION
**Pull request by agkaminski**
JIRA: RTOS-808

## Description

ABI breaking change

## Motivation and Context

Fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/977

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [x] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (ia32-generic-qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Comment by agkaminski** [comment]
Test here: https://github.com/phoenix-rtos/phoenix-rtos-project/pull/1050

File: syscalls.c [hunk 1]
```diff
@@ -942,12 +1080,17 @@ ssize_t syscalls_sys_read(char *ustack)
 	GETFROMSTACK(ustack, void *, buf, 1);
 	GETFROMSTACK(ustack, size_t, nbyte, 2);
 
+	if ((buf != NULL) && (nbyte != 0) && (vm_mapBelongs(proc->mapp, buf, nbyte, proc) < 0)) {
```
Comment (lukileczo) [inline syscalls.c:1083]: I guess not exactly in scope of this PR but with unix socket `buf` is used without any additional checks, so NULL would still cause PF

Comment (agkaminski) [reply syscalls.c:1083]: I only accept `buf == NULL` when `nbyte == 0`, `read()` should be resistant to such parameters (and I think it is indeed)

Comment (agkaminski) [reply syscalls.c:1083]: Oh, ok, this condition is not ideal, will fix

Comment (agkaminski) [reply syscalls.c:1087]: Yeah, this condition wasn't ideal, added a second one to patch it

**Comment by agkaminski** [comment]
@ziemleszcz Could you please check if I had correctly guessed which fields are optional in sockets/inet?

**Review by lukileczo — APPROVED** [review]
LGTM

File: syscalls.c [hunk 2]
```diff
@@ -1215,6 +1440,16 @@ int syscalls_sys_getsockopt(char *ustack)
 	GETFROMSTACK(ustack, void *, optval, 3);
 	GETFROMSTACK(ustack, socklen_t *, optlen, 4);
 
+	if (optval != NULL) {
+		if (vm_mapBelongs(proc, optlen, sizeof(*optlen)) < 0) {
+			return -EFAULT;
+		}
+
+		if ((*optlen != 0) && (vm_mapBelongs(proc, optval, *optlen) < 0)) {
```
Comment (ziemleszcz) [inline syscalls.c:1448]: Why `*optlen != 0` check?

File: syscalls.c [hunk 3]
```diff
@@ -1197,12 +1413,21 @@ int syscalls_sys_getsockname(char *ustack)
 	GETFROMSTACK(ustack, struct sockaddr *, address, 1);
 	GETFROMSTACK(ustack, socklen_t *, address_len, 2);
 
+	if (vm_mapBelongs(proc, address, sizeof(*address)) < 0) {
```
Comment (ziemleszcz) [inline syscalls.c:1416]: Size of `address` should be passed in `*address_len` (`address_len` must be not NULL). Maybe copy code from `syscalls_sys_getsockopt`.

File: syscalls.c [hunk 4]
```diff
@@ -1183,12 +1390,21 @@ int syscalls_sys_getpeername(char *ustack)
...
```
Comment (ziemleszcz) [inline syscalls.c:1393]: Size of `address` should be passed in `*address_len` (`address_len` must be not NULL). Maybe copy code from `syscalls_sys_getsockopt`.

File: syscalls.c [hunk 5]
```diff
@@ -1157,24 +1354,34 @@ int syscalls_sys_connect(char *ustack)
 	GETFROMSTACK(ustack, const struct sockaddr *, address, 1);
 	GETFROMSTACK(ustack, socklen_t, address_len, 2);
 
+	if (vm_mapBelongs(proc, address, sizeof(*address)) < 0) {
```
Comment (ziemleszcz) [inline syscalls.c:1357]: Change `sizeof(*address)` to `address_len`.

File: syscalls.c [hunk 6]
```diff
@@ -1143,12 +1335,17 @@ int syscalls_sys_bind(char *ustack)
...
```
Comment (ziemleszcz) [inline syscalls.c:1338]: Change `sizeof(*address)` to `address_len`.

File: syscalls.c [hunk 7]
```diff
@@ -1129,12 +1312,21 @@ int syscalls_sys_accept4(char *ustack)
 	GETFROMSTACK(ustack, socklen_t *, address_len, 2);
 	GETFROMSTACK(ustack, int, flags, 3);
 
+	if (vm_mapBelongs(proc, address, sizeof(*address)) < 0) {
```
Comment (ziemleszcz) [inline syscalls.c:1315]: `address` is optional. If not NULL it's size should be passed in `*address_len`.

File: syscalls.c [hunk 8]
```diff
@@ -1113,12 +1287,21 @@ int syscalls_sys_accept(char *ustack)
 	GETFROMSTACK(ustack, struct sockaddr *, address, 1);
 	GETFROMSTACK(ustack, socklen_t *, address_len, 2);
 
+	if (vm_mapBelongs(proc, address, sizeof(*address)) < 0) {
```
Comment (ziemleszcz) [inline syscalls.c:1290]: `address` is optional. If not NULL it's size should be passed in `*address_len`.

**Comment by agkaminski** [comment]
@ziemleszcz thanks, fixed

File: syscalls.c [hunk 9]
```diff
@@ -1281,12 +1549,25 @@ ssize_t syscalls_sys_recvmsg(char *ustack)
 	GETFROMSTACK(ustack, struct msghdr *, msg, 1);
 	GETFROMSTACK(ustack, int, flags, 2);
 
+	if (vm_mapBelongs(proc, msg, sizeof(*msg)) < 0) {
+		return -EFAULT;
+	}
+
+	if (vm_mapBelongs(proc, msg->msg_iov, sizeof(msg->msg_iovlen)) < 0) {
```
Comment (lukileczo) [inline syscalls.c:1556]: From posix:
> `msg_iov` points to an array of `iovec` structures; `msg_iovlen` must be set to the dimension of this array
```suggestion
	if (vm_mapBelongs(proc, msg->msg_iov, sizeof(*msg->msg_iov) * msg->msg_iovlen) < 0) {
```

And there's more:
> In each `iovec` structure, the `iov_base` field specifies a storage area and the `iov_len` field gives its size in bytes. Some of these sizes can be zero.

File: syscalls.c [hunk 10]
```diff
@@ -1281,12 +1549,25 @@ ssize_t syscalls_sys_recvmsg(char *ustack)
...
+		return -EFAULT;
+	}
+
+	if (vm_mapBelongs(proc, msg->msg_control, sizeof(msg->msg_controllen)) < 0) {
```
Comment (lukileczo) [inline syscalls.c:1560]:
```suggestion
	if (vm_mapBelongs(proc, msg->msg_control, msg->msg_controllen) < 0) {
```

File: syscalls.c [hunk 11]
```diff
@@ -1295,6 +1576,18 @@ ssize_t syscalls_sys_sendmsg(char *ustack)
...
```
Comment (lukileczo) [inline syscalls.c:1587]: same as in `recvmsg`

File: syscalls.c [hunk 12]
```diff
@@ -1281,12 +1545,33 @@ ssize_t syscalls_sys_recvmsg(char *ustack)
...
 
+	if (vm_mapBelongs(proc, msg, sizeof(*msg)) < 0) {
+		return -EFAULT;
+	}
+
+	if (vm_mapBelongs(proc, msg->msg_iov, sizeof(*msg->msg_iov) * msg->msg_iovlen) < 0) {
+		return -EFAULT;
+	}
+
+	if ((msg->msg_iov->iov_base != NULL) && (vm_mapBelongs(proc, msg->msg_iov->iov_base, msg->msg_iov->iov_len) < 0)) {
```
Comment (lukileczo) [inline syscalls.c:1556]: This should be a loop over `msg->msg_iov` with `msg->msg_iovlen` iterations (although we don't support `iovlen` > 1 it yet)

File: syscalls.c [hunk 13]
```diff
@@ -1267,12 +1522,21 @@ ssize_t syscalls_sys_sendto(char *ustack)
 	GETFROMSTACK(ustack, const struct sockaddr *, dest_addr, 4);
 	GETFROMSTACK(ustack, socklen_t, dest_len, 5);
 
+	if (vm_mapBelongs(proc, message, length) < 0) {
+		return -EFAULT;
+	}
+
+	if ((dest_addr != NULL) && (vm_mapBelongs(proc, dest_addr, sizeof(dest_len)) < 0)) {
```
Comment (lukileczo) [inline syscalls.c:1529]:
```suggestion
	if ((dest_addr != NULL) && (vm_mapBelongs(proc, dest_addr, dest_len) < 0)) {
```

File: syscalls.c [hunk 14]
```diff
@@ -1295,6 +1580,26 @@ ssize_t syscalls_sys_sendmsg(char *ustack)
 	GETFROMSTACK(ustack, const struct msghdr *, msg, 1);
 	GETFROMSTACK(ustack, int, flags, 2);
 
+	if (vm_mapBelongs(proc, msg, sizeof(*msg)) < 0) {
+		return -EFAULT;
+	}
+
+	if (vm_mapBelongs(proc, msg->msg_iov, sizeof(*msg->msg_iov) * msg->msg_iovlen) < 0) {
```
Comment (ziemleszcz) [inline syscalls.c:1587]: `msg_iov` can be NULL assuming `msg_iovlen` is 0.

File: syscalls.c [hunk 15]
```diff
@@ -1295,6 +1580,26 @@ ssize_t syscalls_sys_sendmsg(char *ustack)
...
+		return -EFAULT;
+	}
+
+	if ((msg->msg_iov->iov_base != NULL) && (vm_mapBelongs(proc, msg->msg_iov->iov_base, msg->msg_iov->iov_len) < 0)) {
```
Comment (ziemleszcz) [inline syscalls.c:1591]: Iterate this check over `msg_iovlen` elements of array.

File: syscalls.c [hunk 16]
```diff
@@ -1281,12 +1545,33 @@ ssize_t syscalls_sys_recvmsg(char *ustack)
 	GETFROMSTACK(ustack, struct msghdr *, msg, 1);
 	GETFROMSTACK(ustack, int, flags, 2);
 
+	if (vm_mapBelongs(proc, msg, sizeof(*msg)) < 0) {
+		return -EFAULT;
+	}
+
+	if (vm_mapBelongs(proc, msg->msg_iov, sizeof(*msg->msg_iov) * msg->msg_iovlen) < 0) {
```
Comment (ziemleszcz) [inline syscalls.c:1552]: `msg_iov` can be NULL assuming `msg_iovlen` is 0.

**Review by ziemleszcz — APPROVED** [review]

**Review by lukileczo — APPROVED** [review]

**Pull request merged**